cmd/git-sync/main.go

```diff
@@ -661,6 +679,10 @@ func main() {
 		go exechookRunner.Run(context.Background())
 	}
 
+	// Setup signal notify channel
+	sigChan := make(chan os.Signal, 1)
+	signal.Ignore(syncSig)
```

Comment:
I'm not sure I understand the need to ignore the signal?

What I don't like is that a signal which arrives JUST BEFORE we go to sleep will be totally ignored, but a signal 0.0001 seconds later will be received.

Why not just leave the signal active?

Reply:
What I found here was that if the signal was not disabled at this point, then the git process would receive the signal (and terminate). I would like to find a better way of dealing with that ... possibly having a signal handler function (goroutine) catch the signal.

Reply:
This might touch on what is happening (https://stackoverflow.com/questions/33165530/prevent-ctrlc-from-interrupting-exec-command-in-golang).

_Perhaps_ a change here: https://github.com/trulede/git-sync/blob/3bf26ac2be804b7337a4951191fd636f2a0a5325/pkg/cmd/cmd.go#L56 , possibly triggered by putting something in the context (ctx) so that its only active when `sigChan` is set

Reply:
The answer at thar StackOverflow seems promising:

```diff
diff --git a/pkg/cmd/cmd.go b/pkg/cmd/cmd.go
index bdadf7d..c869228 100644
--- a/pkg/cmd/cmd.go
+++ b/pkg/cmd/cmd.go
@@ -22,6 +22,7 @@ import (
 	"fmt"
 	"os/exec"
 	"strings"
+	"syscall"
 
 	"github.com/go-logr/logr"
 )
@@ -72,6 +73,9 @@ func runWithStdin(ctx context.Context, log logintf, cwd string, env []string, st
 	cmd.Stdout = outbuf
 	cmd.Stderr = errbuf
 	cmd.Stdin = bytes.NewBufferString(stdin)
+	cmd.SysProcAttr = &syscall.SysProcAttr{
+		Setpgid: true,
+	}
 
 	err := cmd.Run()
 	stdout := strings.TrimSpace(outbuf.String())
```

Can we craft an e2e case that forces this condition?

Reply:
It might be possible to use the slow fetch script (see e2e::sync_fetch_skip_depth_1) and do something like this:

1. Start the sync. Shell script for a "slow fetch" runs in own process, sleeps for 5, then runs the git command.
2. Test sleeps 1 second, to be sure that the shell script has started.
3. Test sends the signal. This signal should _not_ be propagated to the shell script.
4. Wait for the git-sync to complete, wait 4+3 seconds (remaining wait time + normal 3 seconds).

I will try it tomorrow.

Reply:
I'm trying to craft an e2e that forces this but I can't, so far.

Reply:
I'm not buying it.  Here's what I tried.

```
diff --git a/slow_git_clone.sh b/slow_git_clone.sh
index 0724cdb..449e5cd 100755
--- a/slow_git_clone.sh
+++ b/slow_git_clone.sh
@@ -20,5 +20,9 @@ if [ "$1" != "clone" ]; then
   exit $?
 fi
 
-sleep 1.1
+trap "exit 1" HUP
+# shell needs to be foreground for signals to be delivered
+sleep 1.1 &
+wait
+
 git "$@"
diff --git a/test_e2e.sh b/test_e2e.sh
index 0671e2b..bc1fcb7 100755
--- a/test_e2e.sh
+++ b/test_e2e.sh
@@ -1139,6 +1139,33 @@ function e2e::sync_on_signal_number() {
     assert_file_eq "$ROOT"/link/file "$FUNCNAME 2"
 }
 
+##############################################
+# Test sync-on-signal doesn't interrupt git
+##############################################
+function e2e::sync_on_signal_not_interrupt() {
+     # First sync
+    echo "$FUNCNAME 1" > "$REPO"/file
+    git -C "$REPO" commit -qam "$FUNCNAME 1"
+
+    GIT_SYNC \
+        --git="$SLOW_GIT_CLONE" `# will exit on SIGHUP` \
+        --period=100s \
+        --sync-on-signal="SIGHUP" \
+        --repo="file://$REPO" \
+        --branch="$MAIN_BRANCH" \
+        --rev=HEAD \
+        --root="$ROOT" \
+        --link="link" \
+        >> "$1" 2>&1 &
+    sleep 1 # long enough to get into the slow clone
+    CTR=$(docker ps --filter label="git-sync-e2e=$RUNID" --format="{{.ID}}")
+    docker_signal "$CTR" SIGHUP # should kill it
+    wait_for_sync 3
+    assert_link_exists "$ROOT"/link
+    assert_file_exists "$ROOT"/link/file
+    assert_file_eq "$ROOT"/link/file "$FUNCNAME 1"
+}
```

but it never dies - you can futz with the timing so that is GURANTEED and it still does not pass thru git-sync to the git.

Reply:
```keyboard ^C signals the PGRP.  Also SIGTERM seems like a poor choice ;)

Try:
```
rm -rf /tmp/gs; bin/linux_amd64/git-sync -v5 --repo https://github.com/kubernetes/kubernetes --period=6s --root /tmp/gs --sync-on-signal SIGTERM & sleep 2; echo kill; kill -TERM $!; sleep 1; echo jobs; jobs
```

Reply:
Alright. I was not installing the signal handler until after the initial clone/sync. So any attempts at testing on the first clone would not have the desired behaviour ... which confused matters.

Need to check if the behaviour that I previously mentioned is still occuring (manually CTRL-C during the sync). It should, since I would do that after the initial clone. And then ... try again with the E2E test.

Reply:
I had no luck in getting a signal to propagate from git-sync to the slow_git_clone.sh script. My last attempt looked like:

```
if [ "$1" != "clone" ]; then
  git "$@"
  exit $?
fi

set +m
signal_handler() {
  echo XXXXX SIGNAL YYYYY
#  exit
}

trap "signal_handler" HUP INT
sleep 5.1
git "$@"
```

Works perfectly fine from the shell.

Reply:
signals in shell are tricky - during an exec() the shell blocks signals.  sleep is an exec().  Try this:

```
sleep 5.1 &
wait
```

`wait` is a shell builtin.

Reply:
signals in shell are tricky - during an exec() the shell blocks signals.  sleep is an exec().  Try this:

```
sleep 5.1 &
wait
```

`wait` is a shell builtin.

Reply:
> ```diff
> cmd.SysProcAttr = &syscall.SysProcAttr{
> +		Setpgid: true,
> +	}
> ```

With `Setpgid: false`, and running this command:
```
$ bin/linux_amd64/git-sync --repo https://github.com/trulede/raspberry.docker.git --period 6s -v 5 --root /tmp/foo --sync-on-signal="SIGINT"
....
{"logger":"","ts":"2023-02-15 19:00:48.679291","caller":{"file":"main.go","line":794},"level":1,"msg":"next sync","waitTime":"6s"}
{"logger":"","ts":"2023-02-15 19:00:54.684689","caller":{"file":"main.go","line":715},"level":1,"msg":"syncing repo"}
{"logger":"","ts":"2023-02-15 19:00:54.684872","caller":{"file":"main.go","line":949},"level":0,"msg":"sanity-checking git repo","repo":"/tmp/foo"}
{"logger":"","ts":"2023-02-15 19:00:54.685222","caller":{"file":"cmd.go","line":55},"level":5,"msg":"running command","cwd":"/tmp/foo","cmd":"git rev-parse --show-toplevel"}
{"logger":"","ts":"2023-02-15 19:00:54.688410","caller":{"file":"cmd.go","line":55},"level":5,"msg":"running command","cwd":"/tmp/foo","cmd":"git fsck --no-progress --connectivity-only"}
{"logger":"","ts":"2023-02-15 19:00:54.704816","caller":{"file":"main.go","line":930},"level":0,"msg":"root directory is valid","path":"/tmp/foo"}
{"logger":"","ts":"2023-02-15 19:00:54.705000","caller":{"file":"cmd.go","line":55},"level":5,"msg":"running command","cwd":"/tmp/foo","cmd":"git rev-parse HEAD"}
{"logger":"","ts":"2023-02-15 19:00:54.708792","caller":{"file":"cmd.go","line":55},"level":5,"msg":"running command","cwd":"/tmp/foo","cmd":"git ls-remote -q https://github.com/trulede/raspberry.docker.git HEAD"}
^C{"logger":"","ts":"2023-02-15 19:00:54.988749","caller":{"file":"main.go","line":730},"msg":"too many failures, aborting","error":"Run(git ls-remote -q https://github.com/trulede/raspberry.docker.git HEAD): signal: interrupt: { stdout: \"\", stderr: \"\" }","failCount":1}
```
... it is possible to "CTRL-C" just as the next sync begins, killing the underlying command (git ls-remote in this case). As a result git-sync exits.

Setting `Setpgid: true` prevents that behaviour.

Reply:
Again, ^C seems to confuse things.  Try this instead

```
rm -rf /tmp/gs; bin/linux_amd64/git-sync --one-time --repo https://github.com/kubernetes/kubernetes -v 5 --root /tmp/gs --sync-on-signal="SIGINT"
```

I chose this repo because it is big and slow to clone, making timing easier.

Now hit ^C and see that git does get killed:

```
{"logger":"","ts":"2023-02-15 10:43:33.392142","caller":{"file":"cmd.go","line":54},"level":5,"msg":"running command","cwd":"","cmd":"git clone --no-checkout https://github.com/kubernetes/kubernetes /tmp/gs"}
^C{"logger":"","ts":"2023-02-15 10:43:34.768060","caller":{"file":"main.go","line":725},"msg":"too many failures, aborting","error":"Run(git clone --no-checkout https://github.com/kubernetes/kubernetes /tmp/gs): signal: interrupt: { stdout: \"\", stderr: \"Cloning into '/tmp/gs'...\\nfetch-pack: unexpected disconnect while reading sideband packet\" }","failCount":1}
```

But this is not really the representative case.  Now try:

```
{"logger":"","ts":"2023-02-15 10:47:48.278868","caller":{"file":"cmd.go","line":54},"level":5,"msg":"running command","cwd":"","cmd":"git clone --no-checkout https://github.com/kubernetes/kubernetes /tmp/gs"}


SENDING SIGKILL


```

It will eventually complete the sync (it takes well over a minute for me :)

So, I think "hit ^C" is actually doing the right thing - you are telling your shell to "stop this process" which REALLY does mean the whole PGroup.  ^C is more than "send a signal", and doesn't represent how people will use this.